docs/RFCS/freeze.md

```diff
@@ -0,0 +1,90 @@
+- Feature Name: Data/network freeze
+- Status: draft
+- Start Date: 2016-02-18
+- Authors: Ben Darnell
+- RFC PR:
+- Cockroach Issue:
+
+
+# Summary
+
+This RFC outlines the plan for freezing our data formats and network
+protocols.
+
+# Motivation
+
+We currently make backwards-incompatible changes to data formats
+without providing any means to upgrade without data loss. This will
+need to stop before beta for obvious reasons.
+
+# Detailed design
+
+## Freeze plan
+
+The freeze will proceed in several steps.
+
+### Stage 0 (now)
+
+Anything goes; changes to on-disk formats do not require any kind of
+migration path.
+
+### Stage 1: Guaranteed upgrade path
+
+In stage 1, we require backwards-compatibility with data written by
+any previous stage 1 build. It should always be possible to upgrade by
+stopping all of the old nodes and then bringing up the new version.
+It's OK at this stage if old and new versions cannot be run
+concurrently, or if the migration process takes some time (e.g.
+rewriting all data before the node can start up).
+
+### Stage 2: Online upgrades
+
+Beginning in stage 2, we require that any upgrade be able to be
+performed without taking the cluster offline: old and new nodes must
+be able to coexist.
+
+## Affected code
+
+Any code could potentially be affected by the freeze, but areas that
+will deserve special scrutiny include:
+
+* All `.proto` definitions
+* The packages `keys` and `util/encoding`
+* All system tables (defined in `sql/system.go`)
+
+## Migration strategies
+
+It is difficult to come up with a universal migration strategy, since
+different changes will require different approaches (for example,
+`.proto` changes could perhaps be made by rewriting data on disk at
+startup, while changes to key construction may require the change to
+be coordinated in a distributed fashion). Therefore we leave the
+specifics of a migration process until the need arises.
+
+To facilitate future changes, we will introduce version numbers at
+several levels. Initially the behavior around these version numbers
+will be conservative and cross-version communication will be limited.
+That makes these version numbers a blunt instrument to be reserved for
+major changes.
+
+* The on-disk format (via a file that lives outside RocksDB). Servers
+  will refuse to load a database with a higher version number than
+  they understand.
+* The network protocol (via GRPC header). Servers and clients will
```

Comment:
This worked at viewfinder because the centralized server provided a one-way guarantee: the server had to support old clients but the clients never had to support an old server. When the server may be older than the client, the client must choose one of two options:
* Pessimistically speak an older protocol until it has learned the server's version
* Optimistically speak the newest protocol until it gets a "version not supported" error from the server. 

Either way, this can be achieved whether each node advertises a single version or a range of versions; I don't think making this into a range adds much.

Also note that this version number is meant to cover major changes comparable to our move from net/rpc to our homegrown codec to grpc. We wouldn't use this mechanism for the kinds of changes that we made with viewfinder's migration system. Those would be done by adding an in-band version number to the affected request proto (or maybe just `BatchRequest`) and a viewfinder-like migration scheme (when the change is not trivialized by protobuf backwards-compatibility).